apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt

```diff
@@ -266,18 +273,34 @@ private constructor(
           true
         }
       }
-    }.map {
-      when (it) {
-        is OperationResponse -> request.operation
-            .parseJsonResponse(it.payload.jsonReader(), request.executionContext[CustomScalarAdapters]!!)
-            .newBuilder()
-            .requestUuid(request.requestUuid)
-            .build()
-        is OperationError -> throw ApolloNetworkException("Operation error ${request.operation.name()}: ${it.payload}")
-        is NetworkError -> throw ApolloNetworkException("Network error while executing ${request.operation.name()}", it.cause)
+    }.map { response ->
+      when (response) {
+        is OperationResponse -> {
+          val responsePayload = response.payload
+          val requestCustomScalarAdapters = request.executionContext[CustomScalarAdapters]!!
+          val (payload, customScalarAdapters) = if (responsePayload.isDeferred()) {
+            val responseId = response.id ?: ""
+            val deferredJsonMerger = deferredJsonMergers.getOrPut(responseId) { DeferredJsonMerger() }
+            deferredJsonMerger.merge(responsePayload)
+            if (responsePayload["hasNext"] == false) {
+              // Last deferred payload: discard the deferredJsonMerger
+              deferredJsonMergers.remove(responseId)
```

Comment:
Bonus point: you don't need this housekeeping if you have only one deferredJsonMerger. The GC should do all the work.